### ISSUE TITLE
Synthetic proxy responses should contain an identifying header

### ISSUE THREAD

**Issue by adleong** [linkerd/linkerd2#2295]
All responses returned by the proxy are either a proxied response from another server or a synthetic response generated by the proxy (e.g. if the proxy encounters an error and returns a 500 or if the proxy times out a request and returns a 504).

When debugging issues, it can be difficult to know for sure if a response (especially a 500 response with no body) was generated by Linkerd or by a service that Linkerd is proxying to.  If the proxy always adds some kind of identifying header on all synthetic responses, it would make this clear.

The inverse of this is having Linkerd add the Via header to responses that it proxies and this can be useful as well, especially if it is unclear whether a request is transitting a proxy or not (such as when debugging iptables issues).

**Comment by NoraCodes** [comment]
Per RFC 2616: 

> The Via general-header field MUST be used by gateways and proxies to indicate the intermediate protocols and recipients between the user agent and the server on requests, and between the origin server and the client on responses.

So, it seems like the way to go here is to add the Via header when proxying, and not use it when the proxy is the initial source (using the standard Server field).

**Comment by hawkw** [comment]
Is there a security concern about leaking information outside the cluster here due to setting a `Via` header?

**Comment by NoraCodes** [comment]
I think there is minimal security risk to reveal _that_ the cluster is running Linkerd. It's probably not a good idea to include a version number, and we should probably allow it to be turned off, but not adding a Via: header is technically a standards violation and, as noted above, makes debugging more difficult.

**Comment by hawkw** [comment]
Fair enough --- I think you're right that we ought not leak the Linkerd version. Carry on!

**Comment by aryan9600** [comment]
Is this issue still relevant/needs work? Would like to take a stab at it! @hawkw

**Comment by hawkw** [comment]
> Is this issue still relevant/needs work? Would like to take a stab at it! @hawkw

As far as I know, nobody's actively working on this, so you're welcome to take it on! I think implementing what RFC 2616 describes (setting the `Via` header when proxying, and the `Server` header when synthesizing errors) is correct, but it might be worth asking around to make sure there are no issues with Linkerd populating those headers --- @olix0r might have something to add? 

Let us know if you have any questions!

**Comment by aryan9600** [comment]
hey @hawkw, sorry for the late response, the codebase is kinda massive so digging into it was a bit tedious xD
so since we want to add the header based on whether the response was synthesised by the proxy, does it make sense to add the logic [here](https://github.com/linkerd/linkerd2-proxy/blob/00c3394ab36eeff14192e6ae52c41238376daabd/linkerd/app/core/src/errors.rs#L175) since this is where we the logic to generate a proxy response based on an error lives? or should there be a particular `Service` and `Layer` for injecting this header? thanks! :)
